### ISSUE TITLE
Anonymous functions

### ISSUE THREAD

**Issue by Lysxia** [Lysxia/first-class-families#8]
At the moment constructing functions with the existing combinators is still quite a puzzle. There should be an easier way to implement arbitrary anonymous functions.

One idea is to add an "environment" argument to `Eval` that can be referenced/updated by some special fcfs. (That would look very much like apply-style defun.)

**Comment by masaeedu** [comment]
One possibility is to implement `S`/`K`/`I` combinators, which can be to eliminate abstraction and express any term from a lambda calculus.

**Comment by Lysxia** [comment]
That's true, and the library already has sufficiently expressive combinators. My worry was that encoding arbitrary lambda terms into combinators by hand does not seem the most pleasant experience. Another possibility is to always defunctionalize (i.e., name all functions).

**Comment by barci2** [comment]
@Lysxia Another idea: how about lambdas with template haskell?

**Comment by Lysxia** [comment]
Yeah that could work. If you make a package that does it I'd be happy to add a link to it.
